File: 1552-magnetic-force-between-two-balls/1552-magnetic-force-between-two-balls.java
class Solution {
    public int maxDistance(int[] position, int m) {
        Arrays.sort(position);
        int l=1;
        int r=position[position.length-1]-position[0];
        int ans=-1;
        while(l<=r){
            int mid=l+(r-l)/2;
            if(isPos(position,m,mid)){
                l=mid+1;

                ans=mid;
            }else{
                r=mid-1;
            }
        }
        return ans;
    }
    boolean isPos(int a[],int m,int mid){
        
        int last=a[0];
        m--;
        for(int i=1;i<a.length;i++){
            if((a[i]-last)>=mid){
               m--;
                last=a[i];
            }
            if(m==0){
                return true;
            }
        }
        return m==0;
    }
}

File: 1552-magnetic-force-between-two-balls/1552-magnetic-force-between-two-balls.py
class Solution:
    def maxDistance(self, position: List[int], m: int) -> int:
        l=1
        
        position.sort()
        r=position[-1]-position[0]
        ans=-1
        while(l<=r):
            mid=l+(r-l)//2
            if(self.isPos(position,mid,m)):
                l=mid+1
                ans=mid
            else:
                r=mid-1
        return ans
    def isPos(self,a:List[int],m:int,k:int):
        last=a[0]
        k=k-1
        for i in a:
            if (i-last >=m):
                k=k-1
                last=i
            if k==0:
                return True
        return k==0

        